File: leetcode/POTD/2025-02-08/solution.py
import heapq

class NumberContainers:
    def __init__(self):
        self.index_to_number = {}
        self.number_to_indices = {}

    def change(self, index, number):
        if index in self.index_to_number and self.index_to_number[index] == number:
            return
        self.index_to_number[index] = number
        self.number_to_indices.setdefault(number, [])
        heapq.heappush(self.number_to_indices[number], index)

    def find(self, number):
        if number not in self.number_to_indices:
            return -1
        while self.number_to_indices[number] and self.index_to_number.get(self.number_to_indices[number][0]) != number:
            heapq.heappop(self.number_to_indices[number])
        return self.number_to_indices[number][0] if self.number_to_indices[number] else -1
        


# Your NumberContainers object will be instantiated and called as such:
# obj = NumberContainers()
# obj.change(index,number)
# param_2 = obj.find(number)

File: leetcode/POTD/2025-02-08/solution.go
package main

import (
	"container/heap"
)

type MinHeap []int

func (h MinHeap) Len() int           { return len(h) }
func (h MinHeap) Less(i, j int) bool { return h[i] < h[j] }
func (h MinHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *MinHeap) Push(x interface{}) {
	*h = append(*h, x.(int))
}

func (h *MinHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[:n-1]
	return x
}

type NumberContainers struct {
	indexToNumber map[int]int
	numberToHeap  map[int]*MinHeap
}

func Constructor() NumberContainers {
	return NumberContainers{
		indexToNumber: make(map[int]int),
		numberToHeap:  make(map[int]*MinHeap),
	}
}

func (this *NumberContainers) Change(index int, number int) {
	if val, exists := this.indexToNumber[index]; exists && val == number {
		return
	}
	this.indexToNumber[index] = number

	if _, exists := this.numberToHeap[number]; !exists {
		h := &MinHeap{}
		heap.Init(h)
		this.numberToHeap[number] = h
	}

	heap.Push(this.numberToHeap[number], index)
}

func (this *NumberContainers) Find(number int) int {
	if h, exists := this.numberToHeap[number]; exists {
		for h.Len() > 0 {
			top := (*h)[0]
			if this.indexToNumber[top] == number {
				return top
			}
			heap.Pop(h)
		}
	}
	return -1
}

/**
 * Your NumberContainers object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Change(index,number);
 * param_2 := obj.Find(number);
 */
